server/service/oauth.go

```diff
@@ -107,30 +114,49 @@ func (o *OAuth) GetUser(ctx context.Context, code string) (*githubUser, error) {
 		return nil, fmt.Errorf("can't parse github user response: %s", err)
 	}
 
-	user.Role = model.Requester
+	if err = o.setRole(client, &user); err != nil {
+		return nil, err
+	}
+
+	return &user, nil
+}
+
+// setRole sets the Role attribute of the given GitHub user based on the
+// organization and team memberships
+func (o *OAuth) setRole(client *http.Client, user *githubUser) error {
+	defaultRole := model.Requester
+
 	if o.restrictRequesterAccess != "" {
-		err := o.checkAccess(client, o.restrictRequesterAccess, user.Login)
-		if err != nil && err != ErrNoAccess {
-			return nil, err
-		}
-		if err == ErrNoAccess {
-			user.Role = model.Worker
+		if err := o.restrictionChecker.checkAccess(client, o.restrictRequesterAccess, user.Login); err == nil {
+			user.Role = model.Requester
+			return nil
+		} else if err != ErrNoAccess {
+			return err
 		}
+
+		defaultRole = model.Worker
 	}
 
-	if o.restrictAccess != "" && (user.Role == model.Worker || o.restrictRequesterAccess == "") {
-		if err := o.checkAccess(client, o.restrictAccess, user.Login); err != nil {
-			return nil, err
+	if o.restrictAccess != "" {
+		if err := o.restrictionChecker.checkAccess(client, o.restrictAccess, user.Login); err != nil {
+			return err
 		}
+
+		user.Role = defaultRole
```

Comment:
I think this and the following one are redundant; not necessarily because they're repeated outside https://github.com/src-d/code-annotation/pull/108/files#diff-96702da8eec3f188048d4e6658fd4494R149

Reply:
Yeah, I just copied your code as suggested here https://github.com/src-d/code-annotation/pull/103#issuecomment-365614878

I can remove these lines if you tell me to, I though you left them intentionally to make it more legible.

Reply:
mine was only compilable pseudocode :wink: but whatever decision you take will be fine for me :dancer: